message.go

```diff
@@ -13,11 +13,21 @@ type message struct {
 	decodeSlots     []interface{}
 }
 
+func decodeIntoMessage(dec decoder, m *message) error {
+	for _, s := range m.decodeSlots {
+		if err := decodeMessage(dec, m, s); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 func decodeMessage(dec decoder, m *message, i interface{}) error {
 	err := dec.Decode(i)
-	if err == nil {
-		m.remainingFields--
-	}
+	// TODO need to verify whether codec.Decode pulls from the reader if
+	// the decode fails, or whether the reader stays in the same state
+	// as before the Decode
+	m.remainingFields--
```

Comment:
Yeah, I got a deadlock in the tests due to failed decoding. Then I realized
that the production reader probably can't backtrack and thus it advances
the reader even though the decode failed. Mind you, I only experienced this
thanks to the mock decoder, which definitely advanced the reader no matter
what. I can modify the mock decoder to not advance the reader, but I'm not
sure that that would be accurate.
On Oct 22, 2015 16:22, "Fred Akalin" <notifications@github.com> wrote:

> In message.go
> <https://github.com/keybase/go-framed-msgpack-rpc/pull/14#discussion_r42808073>
> :
>
> >  func decodeMessage(dec decoder, m *message, i interface{}) error {
> >  	err := dec.Decode(i)
> > -	if err == nil {
> > -		m.remainingFields--
> > -	}
> > +	// TODO need to verify whether codec.Decode pulls from the reader if
> > +	// the decode fails, or whether the reader stays in the same state
> > +	// as before the Decode
> > +	m.remainingFields--
>
> did you run into something that caused you to do this? if not, i'd prefer
> leaving it as-is until the TODO is done
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/keybase/go-framed-msgpack-rpc/pull/14/files#r42808073>
> .
>